### ISSUE TITLE
Simplifying User Authentication with React Firebase Hooks

### ISSUE THREAD

**Issue by YA010** [Shaunwei/RealChar#195]
I would like to propose the adoption of the npm package "react-firebase-hooks" (https://www.npmjs.com/package/react-firebase-hooks) for the project. This package provides a convenient and straightforward way to handle user authentication, making the code much easier to read, maintain, and implement.

By utilizing "react-firebase-hooks," we can simplify the process of checking if the user is currently logged in and obtain the user's ID token, among other tasks. Instead of managing state with "useState," we can now use the hook "useAuthState(auth)" to retrieve the user, loading, and error states directly. This approach is more concise and enhances code readability.

Furthermore, the loading and error states are represented as Booleans, which makes it easier to handle different scenarios. Additionally, the user state returns various user information, including the profile picture, email, and tokens, in a JSON format, simplifying data parsing.

I highly recommend integrating "react-firebase-hooks" into the project as it streamlines the user authentication process and contributes to a more efficient and maintainable codebase.

**Comment by pycui** [comment]
Thanks for the suggestion! Are you looking to contribute a PR? If not it's ok and we will look into this

**Comment by YA010** [comment]
> Thanks for the suggestion! Are you looking to contribute a PR? If not it's ok and we will look into this

Yes looking to contribute to the repo, getting a understanding of the project code atm and should have a update soon

**Comment by pycui** [comment]
> > Thanks for the suggestion! Are you looking to contribute a PR? If not it's ok and we will look into this
> 
> Yes looking to contribute to the repo, getting a understanding of the project code atm and should have a update soon

Awesome, thanks. Feel free to let us know if you have any question or thought

**Comment by YA010** [comment]
Update:   got progress but stuck on the send token to server function

Managed to replace both functions
 
```
const signInWithGoogle = async (isLoggedIn, setToken) => {

and const SignIn = ({ isLoggedIn, setToken }) =>
```


with a single function that signs the user in and is easy to check if the user is auth in any  file with 

```
just import useAuthState and auth into your file  and use 
const [user, loading, error] = useAuthState(auth);

 if the loading is true it means the user auth is still loading and  if the user is true it means the user is logged in 
```

Both the sign in and out functionality works but the issue is with the "sendtokentoserver" function 

what is the token  or name of the token being sent to the server  ? is it the "oauthIdToken"


example json response when the user is signed in  
`


        "user": {
        "uid": "long number gets returned here",
        "stsTokenManager": {
            "refreshToken": "very long number gets returned here",
            "accessToken": "very long number gets returned here",
            "expirationTime": long number gets returned here
        },
  
        "apiKey": "long number gets returned here",
     
    "_tokenResponse": {
        "federatedId": "link gets returned here",
 
        "localId": "long number gets returned here",
     
        "idToken": "very long number gets returned here",
      
        "oauthAccessToken": "very long number gets returned here",
        "oauthExpireIn": number gets returned here,
        "refreshToken": "very long number gets returned here ",
        "expiresIn": " number gets returned here",
        "oauthIdToken": "very long number gets returned here",
      
    "operationType": "signIn"
}`

**Comment by pycui** [comment]
Thanks @YA010 , the token should be the idToken above

**Comment by pycui** [comment]
Hi - just want to follow up on this, would love to see your work for this!

**Comment by YA010** [comment]
Hi , i had issues implementing it , its most likely me messing up , the previous code response i sent earlier worked but that required changing so much structure and code that i reversed it 

but if anyone else has any suggestions, solutions, or would like to take a look at the problem here is the the npm module and a quick video on how to do it

```
_quick 7 min guide on how to implement firebase react hooks:  https://www.youtube.com/watch?v=zQyrwxMPm88

npm module : https://www.npmjs.com/package/react-firebase-hooks_
```

The problem I faced was getting an error message saying "hook can't be used in non-function components." I used const [user, loading, error] = useAuthState(auth); to get the user object, including the ID token and other relevant information. Unfortunately, "user" returned null even when the user was logged in.

will send my progress later today
